### PR TITLE
Add threshold to ControlCostArea; bug fix in sum over controls 

### PR DESCRIPTION
**Pull request by HarshBabla99**
Bug fixes: 
- The control drive area is computed as a Riemann sum; however, the `dt` factor was missing. Fixed this bug.
- Fixed a bug caused by [dynamiqs #974](https://github.com/dynamiqs/dynamiqs/pull/974). The new `TimeQArray.prefactor()` method in the parent class now adds a prefactor to all children, including `ConstantTimeQArray`. Previously, `qontrol` checked for the presence or absence of a prefactor attribute, but now I check if a Hamiltonian is a `ConstantTimeQArray`.

Optimizations:
- Avoid computing propagator dim at every call
- Minimize numpy dependence
- Remove redundant line in loss()

Feature:
- Added a threshold for the control area, which is especially useful for bosonic control. When optimizing a pulse on a truncated oscillator, this helps prevent displacements beyond the Hilbert space cutoff.

### CODE REVIEW

File: qontrol/cost.py [hunk 1]
```diff
@@ -433,8 +435,8 @@ def evaluate_controls(
         self, result: SolveResult, H: TimeQArray, func: callable
     ) -> Array:
         def _evaluate_at_tsave(_H: TimeQArray) -> Array:
-            if hasattr(_H, 'prefactor'):
-                return jnp.sum(func(vmap(_H.prefactor)(result.tsave)))
+            if not isinstance(_H, ConstantTimeQArray):
+                return jnp.sum(func(_H.prefactor(result.tsave)))
```
Comment (dkweiss31) [inline qontrol/cost.py]: I see that you can now call prefactor with an array making the vmap unnecesary, nice! (I don't recall if that was always the case...)

File: qontrol/cost.py [hunk 2]
```diff
@@ -458,22 +460,24 @@ def __call__(
...
     def __call__(
         self,
         result: SolveResult,
         H: TimeQArray,
         parameters: dict | Array,  # noqa ARG002
     ) -> tuple[tuple[Array, Array]]:
-        cost = jnp.abs(self.evaluate_controls(result, H, lambda x: x))
+        dt = result.tsave[1]-result.tsave[0]
+        control_area = self.evaluate_controls(result, H, lambda x: x*dt)
+        cost = self.cost_multiplier * relu(jnp.abs(control_area)-self.threshold)
```
Comment (dkweiss31) [inline qontrol/cost.py]: Shouldn't we make the same change in ControlCostNorm? Maybe this logic wants to be in the ControlCost _evaluate_at_tsave function?

File: qontrol/optimize.py [hunk 3]
```diff
@@ -6,7 +6,6 @@
 import dynamiqs as dq
 import jax
 import jax.numpy as jnp
-import numpy as np
```
Comment (dkweiss31) [inline qontrol/optimize.py]: I think you'll want to undo the changes to get rid of numpy. You'll notice that all of the calls to numpy are outside of jitted regions. Calls to jax or jax.numpy outside of jitted regions just slow you down AFAIK

File: qontrol/plot.py [hunk 4]
```diff
@@ -2,9 +2,9 @@
 
 from collections.abc import Callable
 
-import jax
 import matplotlib.pyplot as plt
-import numpy as np
```
Comment (dkweiss31) [inline qontrol/plot.py:7]: In this file as well, we shouldn't be calling jnp

**Review by dkweiss31 — CHANGES_REQUESTED** [review]
Looks good so far! See my comments: I think the dt logic wants to be in ControlCost, and you'll want to restore most/all calls to numpy since those calls are outside of jit regions

**Comment by dkweiss31** [comment]
Also for all of these PRs: you'll want to run in terminal

```bash
task format
```
and
```bash
task lint
```
The first call autoformats the code and the second call will optimize some imports but also tell you various things you need to fix (lines can't be longer than 88 characters, remove whitespaces, annotate arguments, etc.)

Notice that this is the reason all of the checks are failing

File: qontrol/optimize.py (hunk 3 above)
Comment (HarshBabla99) [reply qontrol/optimize.py]: I agree this would be the case on a CPU. Is this also true on a GPU? 
I presume numpy commands are always run on the CPU. If we use numpy, the process would be bottlenecked by data transfer back and forth between a GPU and a CPU. 
However, I presume jax.numpy could perform all operations on a GPU directly.
I guess we should compare the slowdown due to jax.numpy versus that due to CPU-GPU data transfer.

File: qontrol/cost.py (hunk 2 above)
Comment (HarshBabla99) [reply qontrol/cost.py]: Of course! That makes sense.

File: qontrol/optimize.py (hunk 3 above)
Comment (dkweiss31) [reply qontrol/optimize.py]: I think the bottleneck as you noticed will be saving data and updating the plots, both of which must happen on a CPU. So there is no harm in using numpy in that context. Again, if the jax code is not in a jitted region, then no matter where you run the code (CPU, GPU, TPU) this will be slower than pure numpy (since the code doesn't get compiled).

Comment (HarshBabla99) [reply qontrol/optimize.py]: That makes sense. Will undo the `np->jnp` changes.

**Comment by HarshBabla99** [comment]
Implemented requested changes.

**Review by dkweiss31 — APPROVED** [review]
LGTM! Thanks @HarshBabla99 !
